impl/src/settings.rs

```diff
@@ -483,77 +468,299 @@ fn validate_settings(
+        .map(|pkg| &pkg.name)
+        .collect::<Vec<&String>>()
+    ));
+  }
+
+  // UNWRAP: Safe due to checks above
+  let settings_value = settings_packages[0].metadata.get("raze").unwrap();
+  RazeSettings::deserialize(settings_value)
+    .with_context(|| format!("Failed to deserialize raze settings: {:?}", settings_value))
+}
+
+/** A struct only to deserialize a Cargo.toml to in search of the legacy syntax for [`RazeSettings`](crate::settings::RazeSettings) */
+#[derive(Debug, Clone, Deserialize)]
+pub struct LegacyCargoToml {
+  pub raze: RazeSettings,
+}
+
+/** Parse [`RazeSettings`](crate::settings::RazeSettings) from a Cargo.toml file using the legacy syntax `[raze]` */
+fn parse_raze_settings_legacy(metadata: &Metadata) -> Result<RazeSettings> {
+  let root_toml = metadata.workspace_root.join("Cargo.toml");
+  let toml_contents = std::fs::read_to_string(&root_toml)?;
+  let data = toml::from_str::<LegacyCargoToml>(&toml_contents).with_context(|| {
+    format!(
+      "Failed to read `[raze]` settings from {}",
+      root_toml.display()
+    )
+  })?;
+  Ok(data.raze)
+}
+
+/** Parses raze settings from the contents of a `Cargo.toml` file */
+fn parse_raze_settings(metadata: Metadata) -> Result<RazeSettings> {
+  // Workspace takes precedence
+  let workspace_level_settigns = metadata.workspace_metadata.get("raze");
+  if let Some(value) = workspace_level_settigns {
+    return parse_raze_settings_workspace(value, &metadata);
+  }
+
+  // Root packages are the next priority
+  if let Some(root_package) = metadata.root_package() {
+    if let Some(value) = root_package.metadata.get("raze") {
+      return parse_raze_settings_root_package(value, root_package);
+    }
+  }
+
+  // Attempt to load legacy settings but do not allow failures to propogate
+  if let Ok(settings) = parse_raze_settings_legacy(&metadata) {
     eprintln!(
-      "WARNING: The top-level [raze] key is deprecated. Please set [package.metadata.raze] \
-       instead."
+      "WARNING: The top-level `[raze]` key is deprecated. Please set `[workspace.metadata.raze]` or \
+       `[package.metadata.raze]` instead."
     );
-
-    return Ok(raze);
+    return Ok(settings);
   }
 
-  return Err(RazeError::Generic(
-    "Cargo.toml has no `raze`, `workspace.metadata.raze` or `package.metadata.raze` field".into(),
-  ));
+  // Finally check any package for settings
+  parse_raze_settings_any_package(&metadata)
+}
+
+struct SettingsMetadataFetcher {
+  pub cargo_bin_path: PathBuf,
 }
 
-pub fn load_settings<T: AsRef<Path>>(cargo_toml_path: T) -> Result<RazeSettings, RazeError> {
-  let path = cargo_toml_path.as_ref();
-  let mut toml = match File::open(path) {
-    Ok(handle) => handle,
-    Err(err) => {
-      return Err(RazeError::Generic(err.to_string()));
-    },
+impl MetadataFetcher for SettingsMetadataFetcher {
+  fn fetch_metadata(&self, working_dir: &Path, _include_deps: bool) -> Result<Metadata> {
+    // This fetch should not require network access.
```

Comment:
What does "should not" mean here? I feel like this should be more definitive (e.g. "does not").

Any reason to be unsure?

Reply:
I've tested this from a clean cache and it works as expected but was unsure if there were any edge cases to test as well. Updated the wording here.